### PR TITLE
Encryption: Provide better alternative using libsodium for encryption

### PR DESCRIPTION
**Pull request by chrisbra**
After the issues #638 and #639 I thought about exploring alternatives to
make the crypt code more modern and robust. So for the last couple of
weeks and spend my spare time exploring things and improving my crypt
skills ;)

One of the mentioned alternatives was
[libsodium](https://libsodium.gitbook.io/doc/), which is a
library providing encryption, hashing and other cryptographic features
for end-users. Fortunately it is cross plattform so can be used across a
variety of platforms.

This is currently work in progress, it compiles and seems to work for
me, but is not yet considered stable. Also, I am currently not sure, the
changes to memline/memfile are actually correct. I don't really that
low-level code of the data blocks :/

The reason I am posting it here is, that I'd like to get feedback about
the current scope.

In addition I am not so sure, how to properly hook into encrypting the
blocks for swapfiles and undo-files. For stream-encoding, libsodium
needs to know, when a stream of messages is completed and adds an
additional tag for verification. However given the current design of the
block-writing code, I am not sure how to detect whether the current call
for encrypting will be the last one. So I would like to have some
guidance here.

For now, I have just sprinkled a couple of static TRUE and FALSE for the
swap file encryption and for undo-file encryption I just disabled
complete encryption.

Also the current read-file and write-file end up using different block
sizes, so I made some slight changes to the read-file code, to ensure,
that blocks of 8K will be read to make sure that the crypt layer can read
the same block size when reading as when writing. Not sure if there is a
better way to achieve this.

And of course, I'd like to get some CI builds done :)

I'll have a look at the windows changes, once I get my windows build
system ready. Not sure how much work this requires.

Todo:

- [x] fix error numbers
- [x] write tests
- [x] make changes to the windows build system
- [ ] enable complete undofile encryption (?)

### CODE REVIEW

**Comment by chrisbra** [comment]
Oh and for development I have used `libsodium-dev` version 1.0.18, which is from january 2019 and readily available in my ubuntu box.

**Comment by brammool** [comment]
Christian wrote:

> After the issues #638 and #639 I thought about exploring alternatives to
> make the crypt code more modern and robust. So for the last couple of
> weeks and spend my spare time exploring things and improving my crypt
> skills ;)
> 
> One of the mentioned alternatives was
> [libsodium](https://libsodium.gitbook.io/doc/), which is a
> library providing encryption, hashing and other cryptographic features
> for end-users. Fortunately it is cross plattform so can be used across a
> variety of platforms.
> 
> This is currently work in progress, it compiles and seems to work for
> me, but is not yet considered stable. Also, I am currently not sure, the
> changes to memline/memfile are actually correct. I don&#39;t really that
> low-level code of the data blocks :/
> 
> The reason I am posting it here is, that I&#39;d like to get feedback about
> the current scope.
> 
> In addition I am not so sure, how to properly hook into encrypting the
> blocks for swapfiles and undo-files. For stream-encoding, libsodium
> needs to know, when a stream of messages is completed and adds an
> additional tag for verification. However given the current design of the
> block-writing code, I am not sure how to detect whether the current call
> for encrypting will be the last one. So I would like to have some
> guidance here.
> 
> For now, I have just sprinkled a couple of static TRUE and FALSE for the
> swap file encryption and for undo-file encryption I just disabled
> complete encryption.
> 
> Also the current read-file and write-file end up using different block
> sizes, so I made some slight changes to the read-file code, to ensure,
> that blocks of 8K will be read to make sure that the crypt layer can read
> the same block size when reading as when writing. Not sure if there is a
> better way to achieve this.
> 
> And of course, I&#39;d like to get some CI builds done :)
> 
> I&#39;ll have a look at the windows changes, once I get my windows build
> system ready. Not sure how much work this requires.
> 
> Todo:
> 
> - [ ] fix error numbers
> - [ ] write tests
> - [ ] make changes to the windows build system
> - [ ] enable complete undofile encryption (?)


Thanks for taking the time to dive into this.

The swap file implementation really expects fixed size blocks.  Changing
that is going to be very complicated.  How about disabling the swap file
when using this type of encryption?  This also avoids making the
encryption weaker, since each block is encrypted separately.

I'm afraid I won't have time to look into the details.  We should have
several people check that the way the crypt library is used doesn't open
up any holes or weakens the crypt mechanism.



-- 
The average life of an organization chart is six months.  You can safely
ignore any order from your boss that would take six months to complete.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by atoponce** [comment]
I see the Blowfish -> XChaCha20-Poly1305 change fixing #638, but I don't see a KDF replacement for `sha256_key()` in blowfish.c fixing #639. You've imported libsodium, so you can [take advantage of the `crypto_pwhash_*` API](https://libsodium.gitbook.io/doc/password_hashing) which uses Argon2id, or `crypto_pwhash_scryptsalsa208sha256_*` which uses Scrypt.

**Comment by chrisbra** [comment]
Yes, blowfish is kept separate. But the `XChaCha20-Poly1305` uses the `crypt_pwhash` API and it also uses the `randombytes` API for the salt and the seed. 

I did not want to touch the blowfish code for now, as I don't know, if this is backwards compatible (e.g. if this will can still decrypt existing blowfish encrypted files (swap files, undofile, files))

**Comment by chrisbra** [comment]
okay, let's disable the swapfile encryption. I also made a couple of other small adjustments (such as proper key derivation). Concrete further suggestions welcome!

**Comment by brammool** [comment]
Christian wrote:

> okay, let's disable the swapfile encryption. I also made a couple of
> other small adjustments (such as proper key derivation). Concrete
> further suggestions welcome!

I think you mean "disable the swapfile".  When using strong encryption
writing the swapfile (accidentally) leaks the text, thus I think we
should make it impossible to enable the swapfile.

The swapfile is temporary, thus some people may think it's OK.  But
these days with SSD and other types of disks deleting a file doesn't
mean the text is gone.  Even reformatting the disk may not help.

-- 
A)bort, R)etry, B)ang it with a large hammer

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -2384,8 +2384,18 @@ A jump table for the options with a short description can be found at |Q_op|.
 			you write the file the encrypted bytes will be
 			different.  The whole undo file is encrypted, not just
 			the pieces of text.
-
-	You should use "blowfish2", also to re-encrypt older files.
+	   xchacha20	XChaCha20 Cipher with Poly1305 Message Authentication
+			Code.  Medium strong till strong encryption.
+			Encryption is provided by the libsodium library, so it
+			requires Vim to be build with |+sodium|
```
Comment (espadrine) [inline runtime/doc/options.txt]:
```suggestion
			requires Vim to be built with |+sodium|
```

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -2384,8 +2384,18 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+			It adds a seed and a message authentication code (MAC)
+			to the file.  Only the text in the undo-file is
+			encrypted. This needs at least a Vim 8.XXXX to read
+			such a file.  Encryption of Swapfiles not supported,
+			therefore swap files will be disabled.
+			Currently experiemental.
```
Comment (espadrine) [inline runtime/doc/options.txt]:
```suggestion
			Currently experimental.
```

File: runtime/doc/options.txt (hunk 1 above)
Comment (chrisbra) [reply runtime/doc/options.txt]: thanks!

File: runtime/doc/options.txt (hunk 2 above)
Comment (chrisbra) [reply runtime/doc/options.txt]: will be fixed on next push

**Comment by chrisbra** [comment]
> I think you mean "disable the swapfile".  When using strong encryption

Yes of course.

> writing the swapfile (accidentally) leaks the text, thus I think we
> should make it impossible to enable the swapfile.

I fixed a couple of more problems to make sure Vim closes the swapfile
and won't try to encrypt it. I also fixed a couple of edge cases that come from 
the fact, that since the new crypt code allocates a new buffer, converting
could cause a buffer-overflow. Also properly free the allocated buffer.

> The swapfile is temporary, thus some people may think it's OK.  But
> these days with SSD and other types of disks deleting a file doesn't
> mean the text is gone.  Even reformatting the disk may not help.

I also disabled the undofile. I thought it worked before, but when writing the tests,
I got a couple of overflows, most likely, because the size was too small.

I added some tests to ensure the buffer size for en/decoding is large enough.

Tests should be there in a minute , once I push. They probably won't yet run,
I need to make changes to the CI, so that vim is build with libsodium.

**Comment by brammool** [comment]
Christian wrote:

> > I think you mean "disable the swapfile".  When using strong encryption
> 
> Yes of course.
> 
> > writing the swapfile (accidentally) leaks the text, thus I think we
> > should make it impossible to enable the swapfile.
> 
> I fixed a couple of more problems to make sure Vim closes the swapfile
> and won't try to encrypt it. I also fixed a couple of edge cases that
> come from the fact, that since the new crypt code allocates a new
> buffer, converting could cause a buffer-overflow. Also properly free
> the allocated buffer.
> 
> > The swapfile is temporary, thus some people may think it's OK.  But
> > these days with SSD and other types of disks deleting a file doesn't
> > mean the text is gone.  Even reformatting the disk may not help.
> 
> I also disabled the undofile. I thought it worked before, but when
> writing the tests, I got a couple of overflows, most likely, because
> the size was too small.
> 
> I added some tests to ensure the buffer size for en/decoding is large enough.
> 
> Tests should be there in a minute , once I push. They probably won't yet run,
> I need to make changes to the CI, so that vim is build with libsodium.

Disabling the undo file for now is OK, but we should really make this
work at some point.  I use the undo file all the time, and rely on being
able to undo changes after abandoning Vim and editing a file again.
It would be unexpected that this fails only because a file is encrypted.

It may require using a different file format for the undo file.

-- 
hundred-and-one symptoms of being an internet addict:
6. You refuse to go to a vacation spot with no electricity and no phone lines.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
> It may require using a different file format for the undo file.

Yes, lets leave that for later. Personally, I think it may make sense to disable undo files for sensitive data.

I have also added some slight changes to the MSVC and MSYS Makefiles. I cannot test the MSVC build however. The MSYS build works however.

**Comment by brammool** [comment]
The change in eval.txt uses spaces instead of tabs.

Is this now good enough to include?  I think we should mark this "experimental", so that users know that the encoding may still change and files might not decrypt with a later version of Vim. This will give users some time to try it out and hopefully some crypt experts can check if we are using the crypt method properly.

**Comment by chrisbra** [comment]
> The change in eval.txt uses spaces instead of tabs.

Fixed that.

> Is this now good enough to include? I think we should mark this "experimental", so that users know that the encoding may still change and files might not decrypt with a later version of Vim. This will give users some time to try it out and hopefully some crypt experts can check if we are using the crypt method properly.

Yes, should be good now. I made a couple of minor adjustments (added proper error numbers) and adjusted the documentation a bit. I think it is good to be tried out now.

**Comment by brammool** [comment]
> > The change in eval.txt uses spaces instead of tabs.
> 
> Fixed that.
> 
> > Is this now good enough to include? I think we should mark this "experimental", so that users know that the encoding may still change and files might not decrypt with a later version of Vim. This will give users some time to try it out and hopefully some crypt experts can check if we are using the crypt method properly.
> 
> Yes, should be good now. I made a couple of minor adjustments (added
> proper error numbers) and adjusted the documentation a bit. I think it
> is good to be tried out now.

I included the patch now, but alread found a couple of problems:

The 'swapfile' option is disabled when 'cryptmethod' is set, but no key
has been entered yet.  The swapfile should only be disabled when
encryption is actually setup, thus the 'key' option is not empty.

When typing a wrong password the error is:
	E1200: Decryption failed: corrupted chunk!

The "corrupted chunk" note is incorrect.  We may not be able to
distinguish between a wrong password (which actually makes a brute-force
attack quite a bit easier) and a corrupted file (or changed binary
format).  Just saying "Decryption failed" should be sufficient.

I hope you can improve this soon.

I wonder why the invalid sample text is so long.

-- 
hundred-and-one symptoms of being an internet addict:
24. You realize there is not a sound in the house and you have no idea where
    your children are.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
Hi Bram, thanks for merging.

>  The 'swapfile' option is disabled when 'cryptmethod' is set, but no key
> has been entered yet.  The swapfile should only be disabled when
> encryption is actually setup, thus the 'key' option is not empty.

Are you talking about manually setting the key? E.g. not after `:X`? It makes sense, 
but manually setting the key should not be done. 

> The "corrupted chunk" note is incorrect.  We may not be able to
distinguish between a wrong password 

Okay makes sense, will change. 


> I wonder why the invalid sample text is so long.

I was trying to get a file that is written in 2 chunks, so 16384 K. I made some mistakes, but thought it would be a good test sample nevertheless.

I'll create a follow-up PR implementing those changes.

**Comment by vim-ml** [comment]
Hi Christian,

On Sun, Jun 20, 2021 at 9:34 AM Christian Brabandt <
***@***.***> wrote:

> Hi Bram, thanks for merging.
>
> The 'swapfile' option is disabled when 'cryptmethod' is set, but no key
> has been entered yet. The swapfile should only be disabled when
> encryption is actually setup, thus the 'key' option is not empty.
>
> Are you talking about manually setting the key? E.g. not after :X? It
> makes sense,
> but manually setting the key should not be done.
>
> The "corrupted chunk" note is incorrect. We may not be able to
> distinguish between a wrong password
>
> Okay makes sense, will change.
>
> I wonder why the invalid sample text is so long.
>
> I was trying to get a file that is written in 2 chunks, so 16384 K. I made
> some mistakes, but thought it would be a good test sample nevertheless.
>
> I'll create a follow-up PR implementing those changes.
>
>
>
After merging these changes, the code coverage has gone down (as expected).
Can you also take a look at adding additional tests for the new crypto code?

Thanks,
Yegappan

**Pull request closed without merging**